File: src/de/saschahlusiak/freebloks/database/HighscoreDB.java
package de.saschahlusiak.freebloks.database;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;

public class HighscoreDB extends FreebloksDB {
	public static final String TABLE = "highscores";

	public static final String HIGHSCORE_ID = "_id";				/* 0 */
	public static final String GAME_MODE_ID = "gamemode";			/* 1 */
	public static final String POINTS_ID = "points";				/* 2 */
	public static final String STONES_LEFT_ID = "stonesleft";		/* 3 */
	public static final String PLAYER_COLOR_ID = "playercolor";	/* 4 */
	public static final String PLACE_ID = "place";					/* 5 */
	public static final String FLAGS_ID = "flags";					/* 6 */

	/* NEVER CHANGE THE INDEX OF COLUMNS FOR BACKWARDS COMPATIBILITY */
	public static final int HIGHSCORE_INDEX = 0;
	public static final int GAME_MODE_INDEX = 1;
	public static final int POINTS_INDEX = 2;
	public static final int STONES_LEFT_INDEX = 3;
	public static final int PLAYER_COLOR_INDEX = 4;
	public static final int PLACE_INDEX = 5;
	public static final int FLAGS_INDEX = 6;
	
	
	public static final int FLAG_IS_PERFECT = 0x01;
	
	
	/* WARNING: The IDs are used in Cursors to query the colums, for compatibility they should NEVER be changed.
	 * Make sure to ONLY append colums. */
	public static final String CREATE_TABLE = "CREATE TABLE " + TABLE + " (" +
			HIGHSCORE_ID + " INTEGER PRIMARY KEY, " +	/* 0 */
			GAME_MODE_ID + " INTEGER, " +				/* 1 */
			POINTS_ID + " INTEGER, " +					/* 2 */
			STONES_LEFT_ID + " INTEGER, " +				/* 3 */
			PLAYER_COLOR_ID + " INTEGER, " +			/* 4 */ 
			PLACE_ID + " INTEGER, " +					/* 5 */ 
			FLAGS_ID + " INTEGER);";					/* 6 */

	public static final String DROP_TABLE = "DROP TABLE IF EXISTS " + TABLE + ";";

	public HighscoreDB(Context context) {
		super(context);
	}

	public long addHighscore(int game_mode,
			int points,
			int stones_left,
			int player_color, 
			int place,
			int flags) {
		ContentValues values = new ContentValues();

		values.put(GAME_MODE_ID, game_mode);
		values.put(POINTS_ID, points);
		values.put(STONES_LEFT_ID, stones_left);
		values.put(PLAYER_COLOR_ID, player_color);
		values.put(PLACE_ID, place);
		values.put(FLAGS_ID, flags);

		return db.insert(TABLE, null, values);
	}

	public boolean clearHighscores() {
		return db.delete(TABLE, null, null) > 0;
	}
	
	public int getTotalNumberOfGames(int game_mode) {
		int count;
		String sql = "SELECT COUNT(*) FROM " + TABLE;
		if (game_mode >= 0)
			sql += " WHERE " + GAME_MODE_ID + " = " + game_mode;
		Cursor c = db.rawQuery(sql, null);
		c.moveToFirst();
		count = c.getInt(0);
		c.close();
		return count;
	}
	
	public int getTotalNumberOfPoints(int game_mode) {
		int count;
		String sql = "SELECT SUM(" + POINTS_ID + ") FROM " + TABLE;
		if (game_mode >= 0)
			sql += " WHERE " + GAME_MODE_ID + " = " + game_mode;
		Cursor c = db.rawQuery(sql, null);
		c.moveToFirst();
		count = c.getInt(0);
		c.close();
		return count;
	}
	
	public int getNumberOfPlace(int game_mode, int place) {
		int count;
		String sql = "SELECT COUNT(*) FROM " + TABLE + " WHERE " + PLACE_ID + " = " + place;
		if (game_mode >= 0)
			sql += " AND " + GAME_MODE_ID + " = " + game_mode;
		Cursor c = db.rawQuery(sql, null);
		c.moveToFirst();
		count = c.getInt(0);
		c.close();
		return count;
	}
	
	public int getNumberOfPerfectGames(int game_mode) {
		int count;
		String sql = "SELECT COUNT(*) FROM " + TABLE + " WHERE " + STONES_LEFT_ID + " = 0";
		if (game_mode >= 0)
			sql += " AND " + GAME_MODE_ID + " = " + game_mode;
		Cursor c = db.rawQuery(sql, null);
		c.moveToFirst();
		count = c.getInt(0);
		c.close();
		return count;
	}
}

File: app/src/main/java/de/saschahlusiak/freebloks/database/HighScoreDB.kt
package de.saschahlusiak.freebloks.database

import android.content.ContentValues
import android.content.Context
import de.saschahlusiak.freebloks.model.GameMode

class HighScoreDB(context: Context) : FreebloksDB(context) {
    fun addHighScore(
        gameMode: GameMode,
        points: Int,
        stonesLeft: Int,
        playerColor: Int,
        place: Int,
        flags: Int
    ) {
        val values = ContentValues().apply {
            put(GAME_MODE_ID, gameMode.ordinal)
            put(POINTS_ID, points)
            put(STONES_LEFT_ID, stonesLeft)
            put(PLAYER_COLOR_ID, playerColor)
            put(PLACE_ID, place)
            put(FLAGS_ID, flags)
        }

        db?.insert(TABLE, null, values)
    }

    fun clearHighScores() {
        db?.delete(TABLE, null, null)
    }

    fun getTotalNumberOfGames(gameMode: GameMode?): Int {
        var sql = "SELECT COUNT(*) FROM $TABLE"
        if (gameMode != null) sql += " WHERE $GAME_MODE_ID = ${gameMode.ordinal}"
        return db?.rawQuery(sql, null)?.use { c ->
            c.moveToFirst()
            c.getInt(0)
        } ?: 0
    }

    fun getTotalNumberOfPoints(game_mode: GameMode?): Int {
        var sql = "SELECT SUM($POINTS_ID) FROM $TABLE"
        if (game_mode != null) sql += " WHERE " + GAME_MODE_ID + " = " + game_mode.ordinal
        return db?.rawQuery(sql, null)?.use { c ->
            c.moveToFirst()
            c.getInt(0)
        } ?: 0
    }

    fun getTotalNumberOfStonesLeft(gameMode: GameMode?): Int {
        var sql = "SELECT SUM($STONES_LEFT_ID) FROM $TABLE"
        if (gameMode != null) sql += " WHERE " + GAME_MODE_ID + " = " + gameMode.ordinal
        return db?.rawQuery(sql, null)?.use { c ->
            c.moveToFirst()
            c.getInt(0)
        } ?: 0
    }

    fun getNumberOfPlace(gameMode: GameMode?, place: Int): Int {
        var sql = "SELECT COUNT(*) FROM $TABLE WHERE $PLACE_ID = $place"
        if (gameMode != null) sql += " AND " + GAME_MODE_ID + " = " + gameMode.ordinal
        return db?.rawQuery(sql, null)?.use { c ->
            c.moveToFirst()
            c.getInt(0)
        } ?: 0
    }

    fun getNumberOfPlace(gameMode: GameMode?, place: Int, color: Int): Int {
        var sql = "SELECT COUNT(*) FROM $TABLE WHERE $PLACE_ID = $place AND $PLAYER_COLOR_ID = $color"
        if (gameMode != null) sql += " AND " + GAME_MODE_ID + " = " + gameMode.ordinal
        return db?.rawQuery(sql, null)?.use { c ->
            c.moveToFirst()
            c.getInt(0)
        } ?: 0
    }

    fun getNumberOfGoodGames(gameMode: GameMode?): Int {
        var sql = "SELECT COUNT(*) FROM $TABLE WHERE $STONES_LEFT_ID = 0"
        if (gameMode != null) sql += " AND " + GAME_MODE_ID + " = " + gameMode.ordinal
        return db?.rawQuery(sql, null)?.use { c ->
            c.moveToFirst()
            c.getInt(0)
        } ?: 0
    }

    fun getNumberOfPerfectGames(game_mode: GameMode?): Int {
        var sql = "SELECT COUNT(*) FROM $TABLE WHERE $STONES_LEFT_ID = 0"
        sql += " AND ($FLAGS_ID&$FLAG_IS_PERFECT)=$FLAG_IS_PERFECT"
        if (game_mode != null) sql += " AND " + GAME_MODE_ID + " = " + game_mode.ordinal
        return db?.rawQuery(sql, null)?.use { c ->
            c.moveToFirst()
            c.getInt(0)
        } ?: 0
    }

    companion object {
        private const val TABLE = "highscores"
        private const val HIGHSCORE_ID = "_id" /* 0 */
        private const val GAME_MODE_ID = "gamemode" /* 1 */
        private const val POINTS_ID = "points" /* 2 */
        private const val STONES_LEFT_ID = "stonesleft" /* 3 */
        private const val PLAYER_COLOR_ID = "playercolor" /* 4 */
        private const val PLACE_ID = "place" /* 5 */
        private const val FLAGS_ID = "flags" /* 6 */

        const val FLAG_IS_PERFECT = 0x01

        /* WARNING: The IDs are used in Cursors to query the colums, for compatibility they should NEVER be changed.
	    * Make sure to ONLY append columns. */
        const val CREATE_TABLE = """
            CREATE TABLE $TABLE (
                $HIGHSCORE_ID INTEGER PRIMARY KEY,
                $GAME_MODE_ID INTEGER, 
                $POINTS_ID INTEGER, 
                $STONES_LEFT_ID INTEGER, 
                $PLAYER_COLOR_ID INTEGER, 
                $PLACE_ID INTEGER, 
                $FLAGS_ID INTEGER)
            ;""" /* 6 */

        const val DROP_TABLE = "DROP TABLE IF EXISTS $TABLE;"
    }
}